networking/eth2/src/main/java/tech/pegasys/artemis/networking/eth2/gossip/topics/validation/BlockValidator.java

```diff
@@ -0,0 +1,122 @@
+package tech.pegasys.artemis.networking.eth2.gossip.topics.validation;
+
+import com.google.common.primitives.UnsignedLong;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.tuweni.bytes.Bytes;
+import tech.pegasys.artemis.bls.BLS;
+import tech.pegasys.artemis.bls.BLSSignature;
+import tech.pegasys.artemis.core.StateTransition;
+import tech.pegasys.artemis.core.exceptions.EpochProcessingException;
+import tech.pegasys.artemis.core.exceptions.SlotProcessingException;
+import tech.pegasys.artemis.datastructures.blocks.SignedBeaconBlock;
+import tech.pegasys.artemis.datastructures.forkchoice.ReadOnlyStore;
+import tech.pegasys.artemis.datastructures.state.BeaconState;
+import tech.pegasys.artemis.datastructures.state.Validator;
+import tech.pegasys.artemis.storage.client.RecentChainData;
+
+import java.util.HashSet;
+import java.util.Set;
+
+import static com.google.common.primitives.UnsignedLong.valueOf;
+import static java.lang.Math.toIntExact;
+import static tech.pegasys.artemis.datastructures.util.BeaconStateUtil.compute_signing_root;
+import static tech.pegasys.artemis.datastructures.util.BeaconStateUtil.get_beacon_proposer_index;
+import static tech.pegasys.artemis.datastructures.util.BeaconStateUtil.get_domain;
+import static tech.pegasys.artemis.statetransition.forkchoice.ForkChoiceUtil.getCurrentSlot;
+import static tech.pegasys.artemis.util.config.Constants.DOMAIN_BEACON_PROPOSER;
+import static tech.pegasys.artemis.util.config.Constants.MAXIMUM_GOSSIP_CLOCK_DISPARITY;
+
+public class BlockValidator {
+  private static final Logger LOG = LogManager.getLogger();
+
+  private final RecentChainData recentChainData;
+  private final StateTransition stateTransition;
+  private final Set<UnsignedLong> receivedValidSignatureBlockSlots = new HashSet<>();
```

Comment:
I think we need to put slot/proposerIndex pairs in this collection ... What if there are multiple forks where we have different valid proposers for the same slot on different forks?

Reply:
Done.